File: shared/types/match.ts
export enum MatchStatus {
  PENDING = "pending",
  WAITING_FOR_PLAYERS = "waiting_for_players",
  IN_PROGRESS = "in_progress",
  COMPLETED = "completed",
  DISPUTED = "disputed",
  CANCELLED = "cancelled",
}

export interface MatchPlayerInfo {
  player_id: string;
  position: number;
  sets_won: number;
  legs_won: number;
  team_id?: string;
  team_position?: number;
  arrived_at_board?: string | null;
  reported_win?: boolean | null;
}

export interface Match {
  id: string;
  tournament_id: string;
  round_number: number;
  match_number: number;
  bracket_position?: string;
  status: MatchStatus;
  started_at?: string;
  completed_at?: string;
  winner_id?: string;
  winner_team_id?: string;
  dartboard_id?: string | null;
  created_at: string;
  updated_at: string;
}

export interface MatchWithPlayers extends Match {
  players: MatchPlayerInfo[];
}

File: backend/schemas/match.py
from pydantic import BaseModel, Field
from typing import Optional, List
from uuid import UUID
from datetime import datetime
from backend.models.match import MatchStatus


class MatchPlayerInfo(BaseModel):
    player_id: UUID
    position: int
    sets_won: int = 0
    legs_won: int = 0
    arrived_at_board: Optional[datetime] = None
    reported_win: Optional[bool] = None
    team_id: Optional[UUID] = None
    team_position: Optional[int] = None


class MatchBase(BaseModel):
    tournament_id: UUID
    round_number: int = Field(..., ge=1)
    match_number: int = Field(..., ge=1)
    bracket_position: Optional[str] = None


class MatchCreate(MatchBase):
    player_ids: List[UUID] = Field(..., min_length=2, max_length=2)


class MatchUpdate(BaseModel):
    status: Optional[MatchStatus] = None
    winner_id: Optional[UUID] = None
    winner_team_id: Optional[UUID] = None


class MatchResponse(MatchBase):
    id: UUID
    status: MatchStatus
    started_at: Optional[datetime]
    completed_at: Optional[datetime]
    winner_id: Optional[UUID]
    winner_team_id: Optional[UUID] = None
    dartboard_id: Optional[UUID] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class MatchWithPlayers(MatchResponse):
    players: List[MatchPlayerInfo]


class MatchPlayerCreate(BaseModel):
    match_id: UUID
    player_id: UUID
    position: int = Field(..., ge=1, le=2)
